backend/typescript/services/implementations/reviewService.ts

```diff
@@ -399,6 +480,266 @@ class ReviewService implements IReviewService {
 
     return result;
   }
+
+  async updateReviews(id: string, entity: ReviewRequestDTO): Promise<void> {
+    try {
+      await this.db.transaction(async (t) => {
+        const reviewToUpdate = await PgReview.findByPk(id, {
+          transaction: t,
+          include: [{ all: true, nested: true }],
+        });
+
+        if (!reviewToUpdate) {
+          throw new Error(`Review id ${id} not found`);
+        }
+
+        // delete tags
+        const newTagIds = entity.tags
+          .filter((tag: TagRequest) => typeof tag.id === "number")
+          .map((tag: TagRequest) => tag.id);
+        const allReviewTags = await PgReviewTag.findAll({
+          where: { review_id: id },
+          transaction: t,
+        });
+        const droppedTags = allReviewTags
+          .map((reviewTag: PgReviewTag) => reviewTag.tag_id)
+          .filter((oldId: number) => !newTagIds.includes(oldId));
+        await Promise.all(
+          droppedTags.map(async (tagId: number) => {
+            const dbTag = await PgTag.findByPk(tagId, { transaction: t });
+            if (!dbTag) {
+              throw new Error(`Tag id ${dbTag} not found`);
+            }
+            reviewToUpdate.$remove("tags", dbTag, { transaction: t });
+          }),
+        );
+
+        // update tags
+        await Promise.all(
+          entity.tags.map(async (tag: TagRequest) => {
+            if (tag.id) {
+              const tagToUpdate = await PgTag.findByPk(tag.id, {
+                transaction: t,
+              });
+              if (!tagToUpdate) {
+                throw new Error(`Tag id ${tag.id} not found`);
+              }
+              await tagToUpdate.update({ name: tag.name });
+            } else {
+              await this.findOrCreateTag(reviewToUpdate, tag, t);
+            }
+          }),
+        );
+
+        // remove books
+        const oldBooks = await PgBook.findAll({ transaction: t });
+        const oldBookIds = oldBooks.map((oldBook: PgBook) => oldBook.id);
+        const newBookIds = entity.books
+          .map((book: BookRequest) => book.id)
+          .filter((bookId: number | undefined) => bookId !== undefined);
+        const droppedBookIds = oldBookIds.filter(
+          (oldId: number) => !newBookIds.includes(oldId),
+        );
+        await Promise.all(
+          droppedBookIds.map(async (bookId: number) => {
+            await PgBookAuthor.destroy({
+              where: {
+                book_id: bookId,
+              },
+              transaction: t,
+            });
+
+            await PgBookPublisher.destroy({
+              where: {
+                book_id: bookId,
+              },
+              transaction: t,
+            });
+
+            await PgBook.destroy({ where: { id: bookId }, transaction: t });
+          }),
+        );
+
+        // update books
+        await Promise.all(
```

Comment:
same as this promise.all(). It might not be needed?

Reply:
just check for one and then apply that to all the Promise.all() lol. Idk if promise.all() needs to be used at all if all the promises inside are being awaited